docs/proposals/20190610-machine-states-preboot-bootstrapping.md

```diff
@@ -0,0 +1,446 @@
+    - **Phase [optional]**  _Superseded by Phase_
+        - Type: `*string`
+        - Description: Phase represents the current phase of machine actuation. E.g. Pending, Running, Terminating, Failed etc.
+- **To add**
+    - **Phase**
+        - Type: `MachinePhaseType`
+        - Description: Phase represents the current state of a Machine. This field is not meant to be used as a source of truth. Controllers working on Machines should always respect the documented transition conditions to determine the state.
+    - **Version [optional]**
+        - Type: ``*string`
+        - Description: Version specifies the current version of Kubernetes running on the corresponding Node. This is meant to be a means of bubbling up status from the Node to the Machine. It is entirely optional, but useful for end-user UX if it’s present.
+    - **BootstrapPhase [optional]**
+        - Type: `ControllerPhaseType`
+        - Description: The state of the bootstrap provider.
+    - **InfrastructurePhase [optional]**
+        - Type: `ControllerPhaseType`
+        - Description: The state of the infrastructure provider.
+
+```go
+type Bootstrap struct
+```
+- **To add**
+    - **ConfigRef [optional]**
+        - Type: `*corev1.TypedLocalObjectReference`
+        - Description: ConfigRef is a reference to a bootstrap provider-specific resource that holds configuration details. The reference is optional to allow users/operators to specify Bootstrap.Data without the need of a controller.
+    - **Data [optional]**
+        - Type: `*string`
+        - Description: Data contains the bootstrap data, such as cloud-init details scripts. If nil, the Machine should remain in the Pending state.
+
+## States and transitions
+The Cluster API Machine Controller is responsible for managing the lifecycle of a Machine, including its state transitions. This is the only controller that should be writing updates to Machines.
+
+---
+### Pending
+```go
+// MachinePhasePending is the first state a Machine is assigned by
+// Cluster API Machine controller after being created.
+MachinePhasePending = MachinePhaseType("pending")
+```
+
+#### Transition Conditions
+- `Machine.Status.Phase` is empty string
+
+#### Expectations
+- When `Machine.Spec.Bootstrap.Data` is:
+    - `<nil>`, expect the field to be set by an external controller.
+    - `""` (empty string), expect the bootstrap step to be ignored.
+    - `"..."` (populated by user or from the bootstrap provider), expect the contents to be used by a bootstrap (pre-Running transition) or infrastructure (Provisioning state) provider.
+
+---
+### Provisioning
+```go
+// MachinePhaseProvisioning is the state when the
+// Machine infrastructure is being created.
+MachinePhaseProvisioning = MachinePhaseType("provisioning")
+```
+
+#### Transition Conditions
+- `Machine.Status.Phase` is set to “Pending”
```

Comment:
I don't think we want to specify existing phase settings as conditions for transitioning. That doesn't match our caveat that nothing should be looking at the phase to decide what to do, and forces us into a one-thing-at-a-time implementation instead of allowing controllers to operate in parallel. At some point we talked about the phase settings happening as a result of the transition, rather than a condition. Maybe that was not reflected in this version of the document?

In addition to that, the literal values listed in these "transition condition" sections don't match the code examples. I think you could just take out all of the "Phase is set to ..." lines to resolve this.

Reply:
Fixed it, thanks!